src/Altinn.App.Core/Internal/Process/ProcessTasks/Common/ProcessTaskFinalizer.cs

```diff
@@ -54,58 +47,107 @@ public async Task Finalize(string taskId, Instance instance)
         ApplicationMetadata applicationMetadata = await _appMetadata.GetApplicationMetadata();
         List<DataType> connectedDataTypes = applicationMetadata.DataTypes.FindAll(dt => dt.TaskId == taskId);
 
-        await RunRemoveFieldsInModelOnTaskComplete(instance, connectedDataTypes);
+        var dataAccessor = new CachedInstanceDataAccessor(instance, _dataClient, _appMetadata, _appModel);
+        var changedDataElements = await RunRemoveFieldsInModelOnTaskComplete(
+            instance,
+            dataAccessor,
+            taskId,
+            connectedDataTypes,
+            language: null
+        );
+
+        // Save changes to the data elements with app logic that was changed.
+        await Task.WhenAll(
+            changedDataElements.Select(async dataElement =>
+            {
+                var data = await dataAccessor.GetData(dataElement);
+                return _dataClient.UpdateData(
+                    data,
+                    Guid.Parse(instance.Id.Split('/')[1]),
+                    data.GetType(),
+                    instance.Org,
+                    instance.AppId.Split('/')[1],
+                    int.Parse(instance.InstanceOwner.PartyId, CultureInfo.InvariantCulture),
+                    Guid.Parse(dataElement.Id)
+                );
+            })
+        );
```

Comment:
I think the other places `CachedInstanceDataAccessor` is used, the cache is updated after updates/modifications, but not here, maybe because it isn't needed. If it is true that when using `CachedInstanceDataAccessor` and making data updates, we should always make sure the cache is in sync, can we write the code such that updating data always updates  the cache or vice versa?

Reply:
If you just mutate the data models, the cache will be in sync. The issue is if you add or remove data elements from the instance, or if you want to replace the root data object in the cache.

We actually have such a modification when removing shadow fields (or making a copy of the data element without shadow fields)

https://github.com/Altinn/app-lib-dotnet/blob/450f53d978f52e65c275b8146d4c6ac6ebea1c55/src/Altinn.App.Core/Internal/Process/ProcessTasks/Common/ProcessTaskFinalizer.cs#L157-L185

Maybe `IInstanceDataProcessor` should have apis to mutate Instance.Data and sync with both local state and storage?

Reply:
This has been resolved by splitting the interface into `IInstanceDataAccessor` and `IInstanceDataMutator`. One for validations, and the mutator for actions and data processing.
 CachedInstanceDataAccessor implements both and can be used in both